### PR TITLE
PATCHED_API_KEY integration

### PR DESCRIPTION
**Pull request by CTY-git**

### CODE REVIEW

**Comment by patched-admin** [comment]
In the `CallOpenAI.py` file, a try-except block has been added to catch exceptions when running the "patchflow_instance", logging an error message and exiting the program with code 1 if an exception occurs. Additionally, default URLs for Patched and Patch have been included, the OpenAI API key is no longer a required key, and the initialization process now checks for an OpenAI API key in the inputs or environment variables, as well as a Patched API key. If an API key is not found, a ValueError is raised with instructions on how to obtain and set the API keys. In a Python file, the formatting of the `sorted` function call in the `run` method has been modified by splitting it into multiple lines for clarity and readability.

------

<div markdown="1">

* File changed: [patchwork/app.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)
  [The code has been modified to include a try-except block that catches exceptions when running the "patchflow_instance". If an exception occurs, it logs an error message and exits the program with code 1.](https://github.com/patched-codes/patchwork/pull/12/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)

</div>

<div markdown="1">

* File changed: [patchwork/steps/CallOpenAI/CallOpenAI.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-1a65c740612bf992363dad93d1e86d0579a0015d2294ffa15404cd601972c925)
  [The changes in the `CallOpenAI.py` file include adding default URLs for Patched and Patch, removing OpenAI API key as a required key, and updating the initialization process to include checking for an OpenAI API key in the inputs or environment variables, alongside a check for a Patched API key. If an API key is not found, it raises a ValueError with instructions on how to obtain and set the API keys.](https://github.com/patched-codes/patchwork/pull/12/files#diff-1a65c740612bf992363dad93d1e86d0579a0015d2294ffa15404cd601972c925)

</div>

<div markdown="1">

* File changed: [patchwork/steps/ModifyCode/ModifyCode.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-10e9c07de238f7931e667c3d1d277b8b9ba8850bbeced487eb2de77ae6edc278)
  [The diff involves modifying the formatting of the `sorted` function call in the `run` method of a Python file. The function call has been split into multiple lines for clarity and readability.](https://github.com/patched-codes/patchwork/pull/12/files#diff-10e9c07de238f7931e667c3d1d277b8b9ba8850bbeced487eb2de77ae6edc278)

</div>

File: patchwork/steps/CallOpenAI/CallOpenAI.py [hunk 1]
```diff
@@ -10,9 +10,12 @@
 from patchwork.logger import logger
 from patchwork.step import Step
 
+_TOKEN_URL = "https://app.patched.codes/signin"
+_DEFAULT_PATCH_URL = "https://patch-function-ps63srnnsq-de.a.run.app"
```
Comment (codelion) [inline patchwork/steps/CallOpenAI/CallOpenAI.py:14]: Can we avoid exposing the public cloud run URL. We can use something like "https://api.patched.codes/v1/"

**Comment by patched-admin** [comment]
The changes in the code include importing the "traceback" module, adding exception handling to log and display error messages during patchflow execution, modifying the `create_comment()` method to handle `None` path scenarios, updating comment creation logic, file difference retrieval, and `slug` calculation in the `get_pr_by_url()` method. The CallOpenAI.py file now includes handling Patched API key and default Patch Function URL, with validation adjustments raising errors for missing 'openai_api_key'. Additionally, a code snippet was updated to sort code snippets and responses based on start line in reverse order, with the sorting code formatted for readability on multiple lines.

------

<div markdown="1">

* File changed: [patchwork/app.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)
  [A new import statement for the "traceback" module was added. Additionally, an exception handling block was included in the code to catch any exceptions during the execution of the patchflow, log the traceback, and display an error message before exiting the program with an error code.](https://github.com/patched-codes/patchwork/pull/12/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)

</div>

<div markdown="1">

* File changed: [patchwork/common/client/scm.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-a4a0213f577e66fc33444faad6c1991a8c60deb8fc7d5b5527069a66f59804f5)
  [The changes in the diff include altering the arguments of the `create_comment()` method, restructuring the method to handle a scenario where `path` is `None`, modifying the logic for creating comments, updating the method for retrieving file differences, and adjusting the calculation for `slug` in the `get_pr_by_url()` method.](https://github.com/patched-codes/patchwork/pull/12/files#diff-a4a0213f577e66fc33444faad6c1991a8c60deb8fc7d5b5527069a66f59804f5)

</div>

<div markdown="1">

* File changed: [patchwork/steps/CallOpenAI/CallOpenAI.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-1a65c740612bf992363dad93d1e86d0579a0015d2294ffa15404cd601972c925)
  [The code in the CallOpenAI.py file has been modified to include handling of Patched API key and default Patch Function URL, and the validation for 'openai_api_key' has been adjusted to raise errors if the key is not found or missing.](https://github.com/patched-codes/patchwork/pull/12/files#diff-1a65c740612bf992363dad93d1e86d0579a0015d2294ffa15404cd601972c925)

</div>

<div markdown="1">

* File changed: [patchwork/steps/ModifyCode/ModifyCode.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-10e9c07de238f7931e667c3d1d277b8b9ba8850bbeced487eb2de77ae6edc278)
  [In the given diff, a code snippet was updated to sort a list of code snippets and extracted responses based on the start line in reverse order. The change involved formatting the sorting code onto multiple lines for better readability.](https://github.com/patched-codes/patchwork/pull/12/files#diff-10e9c07de238f7931e667c3d1d277b8b9ba8850bbeced487eb2de77ae6edc278)

</div>

**Comment by patched-admin** [comment]
In the `app.py` file, the `cli` function now includes a try-except block to handle exceptions during the execution of `patchflow_instance.run()`, logging error messages and tracebacks before exiting with an error status code. Changes in the scm.py file involve modifications to the create_comment and file_diffs methods in a client module, with new logic for handling a path parameter and retrieving file differences for a Merge Request. The CallOpenAI.py file no longer requires "openai_api_key" as a key, has updated initialization for handling different API keys, a new default Patch URL, and validation checks for missing or required API keys. Additionally, the ModifyCode.py file now has improved formatting with the line sorting operation moved to a separate line for clarity.

------

<div markdown="1">

* File changed: [patchwork/app.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)
  [The code in the `cli` function of the `app.py` file has been updated to include a try-except block for handling exceptions. If an exception occurs during the execution of `patchflow_instance.run()`, the error message and traceback will be logged, and the application will exit with an error status code.](https://github.com/patched-codes/patchwork/pull/12/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)

</div>

<div markdown="1">

* File changed: [patchwork/common/client/scm.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-a4a0213f577e66fc33444faad6c1991a8c60deb8fc7d5b5527069a66f59804f5)
  [The diff in the scm.py file includes changes to the create_comment and file_diffs methods in a client module. The create_comment method now includes additional logic to handle a new path parameter and updates the functionality within the method. Additionally, the file_diffs method includes changes to retrieve and return file differences related to a Merge Request.](https://github.com/patched-codes/patchwork/pull/12/files#diff-a4a0213f577e66fc33444faad6c1991a8c60deb8fc7d5b5527069a66f59804f5)

</div>

<div markdown="1">

* File changed: [patchwork/steps/CallOpenAI/CallOpenAI.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-1a65c740612bf992363dad93d1e86d0579a0015d2294ffa15404cd601972c925)
  [The changes in the CallOpenAI.py file include removing "openai_api_key" from the required keys, modifying the initialization to handle different API keys, introducing a new default Patch URL, and adding validation checks and prompts for missing or required API keys before running the step.](https://github.com/patched-codes/patchwork/pull/12/files#diff-1a65c740612bf992363dad93d1e86d0579a0015d2294ffa15404cd601972c925)

</div>

<div markdown="1">

* File changed: [patchwork/steps/ModifyCode/ModifyCode.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-10e9c07de238f7931e667c3d1d277b8b9ba8850bbeced487eb2de77ae6edc278)
  [The diff in ModifyCode.py file shows a change in formatting where the line sorting operation is now within a separate line for better readability.](https://github.com/patched-codes/patchwork/pull/12/files#diff-10e9c07de238f7931e667c3d1d277b8b9ba8850bbeced487eb2de77ae6edc278)

</div>

**Comment by patched-admin** [comment]
The first change involves adding an import statement for the `traceback` module and improving error handling in the `cli` function by implementing a try-except block to catch exceptions, log the traceback, display the error message, and exit with error code 1. In scm.py, adjustments were made to the create_comment() method to create a note if the path is None, and changes were made to how comments are handled in discussions. Additionally, file_diffs() now handles the latest diff and returns a dictionary of file paths and their differences. The code for CallOpenAI was updated to support a patched API key if an OpenAI key is not provided, raising an error if neither key is given and providing guidance on obtaining and setting the keys. In ModifyCode.py, the diff shows additional formatting for sorting `sorted_list` in the `run` function, with the lambda function now spanning multiple lines for improved readability.

------

<div markdown="1">

* File changed: [patchwork/app.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)
  [A new import statement for the `traceback` module has been added. Error handling has been improved in the `cli` function by adding a try-except block to catch exceptions, log the traceback, display the error message, and exit with error code 1.](https://github.com/patched-codes/patchwork/pull/12/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)

</div>

<div markdown="1">

* File changed: [patchwork/common/client/scm.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-a4a0213f577e66fc33444faad6c1991a8c60deb8fc7d5b5527069a66f59804f5)
  [The diff in scm.py involves changes to the create_comment() method where a condition is added to create a note if the path is None, along with adjustments to how comments are handled in discussions. Additionally, changes were made to the file_diffs() method to handle the latest diff and return a dictionary of file paths and their corresponding differences.](https://github.com/patched-codes/patchwork/pull/12/files#diff-a4a0213f577e66fc33444faad6c1991a8c60deb8fc7d5b5527069a66f59804f5)

</div>

<div markdown="1">

* File changed: [patchwork/steps/CallOpenAI/CallOpenAI.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-1a65c740612bf992363dad93d1e86d0579a0015d2294ffa15404cd601972c925)
  [The code for CallOpenAI has been updated to support a patched API key in addition to the OpenAI API key. If the OpenAI key is not provided, it falls back to the Patched API key with the corresponding base URL. An error is raised if neither key is provided and guidance is provided for obtaining and setting the keys.](https://github.com/patched-codes/patchwork/pull/12/files#diff-1a65c740612bf992363dad93d1e86d0579a0015d2294ffa15404cd601972c925)

</div>

<div markdown="1">

* File changed: [patchwork/steps/ModifyCode/ModifyCode.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-10e9c07de238f7931e667c3d1d277b8b9ba8850bbeced487eb2de77ae6edc278)
  [The diff shows a change in the code where additional formatting was introduced for sorting `sorted_list` in the `run` function of the `ModifyCode.py` file, with the lambda function now being formatted across multiple lines for better readability.](https://github.com/patched-codes/patchwork/pull/12/files#diff-10e9c07de238f7931e667c3d1d277b8b9ba8850bbeced487eb2de77ae6edc278)

</div>

**Review by codelion — APPROVED** [review]
LGTM!

**Comment by patched-admin** [comment]
The code changes involve importing the traceback module, adding exception handling, and printing traceback information and error messages in case of an exception. Additionally, modifications to the 'create_comment' method in the 'scm.py' file of the patchwork module include setting a 'note' based on 'path,' handling file differences, and changing the logic for extracting 'slug' from a PR URL. In the 'CallOpenAI.py' file, the required_keys attribute in the CallOpenAI class is updated, constants for _TOKEN_URL and _DEFAULT_PATCH_URL are introduced, and input API key handling is modified with validation checks. A new function `transform_sarif_results` replaces existing code in the `ExtractCode` class, generating code context based on parameters. Other changes involve formatting in the `ModifyCode.py` file for better readability and adding `filter_by_extension` function in the `ReadPRDiffs` class to ignore specific file extensions during processing.

------

<div markdown="1">

* File changed: [patchwork/app.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)
  [The code has been updated to include importing the traceback module, adding exception handling around the instantiation and execution of a class instance, printing the traceback information and the error message if an exception occurs, and exiting the program with code 1 in case of an error.](https://github.com/patched-codes/patchwork/pull/12/files#diff-839e90b808d34e4cf447eff0896161788ccfc6e1f2970be2e551b64ba413a503)

</div>

<div markdown="1">

* File changed: [patchwork/common/client/scm.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-a4a0213f577e66fc33444faad6c1991a8c60deb8fc7d5b5527069a66f59804f5)
  [The changes in the diff involve modifications to the 'create_comment' method in the 'scm.py' file of the patchwork module. The modifications include setting a 'note' based on the value of 'path' and handling file differences based on the latest diff created. Additionally, there is a change in the logic for extracting the 'slug' from a PR URL.](https://github.com/patched-codes/patchwork/pull/12/files#diff-a4a0213f577e66fc33444faad6c1991a8c60deb8fc7d5b5527069a66f59804f5)

</div>

<div markdown="1">

* File changed: [patchwork/steps/CallOpenAI/CallOpenAI.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-1a65c740612bf992363dad93d1e86d0579a0015d2294ffa15404cd601972c925)
  [The patch in the file `CallOpenAI.py` updates the required_keys attribute in the CallOpenAI class by removing "openai_api_key." It introduces new constants for _TOKEN_URL and _DEFAULT_PATCH_URL. Additionally, it modifies the handling of input API keys and adds validation checks for the existence of the required keys before proceeding with the execution.](https://github.com/patched-codes/patchwork/pull/12/files#diff-1a65c740612bf992363dad93d1e86d0579a0015d2294ffa15404cd601972c925)

</div>

<div markdown="1">

* File changed: [patchwork/steps/ExtractCode/ExtractCode.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-065957bbe7fa12ec637b493e502a0c093c283897108bed8db583c468989a877b)
  [The diff adds a new function `transform_sarif_results` that processes SARIF data to extract code context based on various parameters like context length and vulnerability limit. This new function replaces the existing code that performed a similar task within the `ExtractCode` class.](https://github.com/patched-codes/patchwork/pull/12/files#diff-065957bbe7fa12ec637b493e502a0c093c283897108bed8db583c468989a877b)

</div>

<div markdown="1">

* File changed: [patchwork/steps/ModifyCode/ModifyCode.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-10e9c07de238f7931e667c3d1d277b8b9ba8850bbeced487eb2de77ae6edc278)
  [The diff includes a change in formatting in the `ModifyCode.py` file where a sorted list is generated, and now the `sorted` function is written with the lambda function on a new line with the parameters properly aligned for better readability.](https://github.com/patched-codes/patchwork/pull/12/files#diff-10e9c07de238f7931e667c3d1d277b8b9ba8850bbeced487eb2de77ae6edc278)

</div>

<div markdown="1">

* File changed: [patchwork/steps/ReadPRDiffs/ReadPRDiffs.py](https://github.com/patched-codes/patchwork/pull/12/files#diff-98bf6d1177ab0c2533443d9ff06b9e82fee1e552e7ac1a1196ee657492f35a61)
  [This diff introduces a new function called `filter_by_extension` in the `ReadPRDiffs` class of the ReadPRDiffs.py file. It also adds a list named `_IGNORED_EXTENSIONS` with various file extensions that are to be ignored during processing. If a file path matches any of these extensions, it will be skipped during the operation.](https://github.com/patched-codes/patchwork/pull/12/files#diff-98bf6d1177ab0c2533443d9ff06b9e82fee1e552e7ac1a1196ee657492f35a61)

</div>

**Pull request merged**